aip/client-libraries/4221.md

```diff
@@ -0,0 +1,80 @@
+---
+aip:
+  id: 4221
+  scope: client-libraries
+  state: draft
+  created: 2019-06-17
+permalink: /client-libraries/4221
+redirect_from:
+  - /4221
+---
+
+# Client-side retry
+
+Many APIs have error modes that a client should retry. These error modes vary
+across APIs, leaving users to hand-write retry logic around client libraries
+based on available documentation.
+
+## Guidance
+
+Client libraries **may** provide automatic client-side retry logic. Client
+libraries with automatic client-side retry logic **should** provide mechanisms
+(e.g. client initialization options) for users to configure it.
+
+Client libraries for API systems that support remotely-resolved client retry 
+configuration **should** respect the remotely-resolved configuration. However,
+user configuration via the aforementioned mechanisms **must** be respected.
+
+Client library generators implementing this feature **must** accept a retry
+configuration. This retry configuration **may** be supplied via a `protoc`
+plugin option. In the absence of a given retry configuration, client library
+generators **should not** generate a default retry configuration.
```

Comment:
Even retrying UNAVAILABLE and ABORTED on GETs?

Reply:
My first thought was to not suggest this in the spirit of gRPC's "something or nothing" type of approach. For example, with regular gRPC clients, if no remote gRPC ServiceConfig is resolved, clients won't retry *anything* by default, even methods/codes that *look* safe.

I am happy to come back on this though and include what you've suggested. WDYT?

Reply:
What does Jean think?

Reply:
@jadekler WDYT? Should a generator generate a default retry config for UNAVAILABLE & ABORTED on GETs?

Reply:
Offline discussion with @lukesneeringer & @jadekler reached consensus on the original approach: generators should not generate a default retry configuration if a config is not provided.